File: src/Vpiska.Api/Vpiska.Api.Control/Infrastructure/Firebase/FileStorage.fs
namespace Vpiska.Api.Control.Infrastructure.Firebase

open System.IO
open System.Net
open FSharp.Control.Tasks
open Google
open Google.Cloud.Storage.V1

type FirebaseSettings = { BucketName: string }

module FileStorage =
    
    let uploadFile (storageClient: StorageClient) (bucketName: string) (imageId: string) (data: byte[]) (contentType: string) =
        task {
            use stream = new MemoryStream(data)
            let! result = storageClient.UploadObjectAsync(bucketName, imageId, contentType, stream)
            return result.Name
        }
        
    let deleteFile (storageClient: StorageClient) (bucketName: string) (imageId: string) =
        task {
            try
                do! storageClient.DeleteObjectAsync(bucketName, imageId)
                return true
            with
            | :? GoogleApiException as ex ->
                match ex.HttpStatusCode with
                | HttpStatusCode.NotFound -> return false
                | _ -> return raise ex
        }

File: src/Vpiska.Infrastructure/Firebase/FileStorage.cs
using System.IO;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
using Google;
using Google.Cloud.Storage.V1;
using Vpiska.Domain.Common;
using Vpiska.Domain.Interfaces;

namespace Vpiska.Infrastructure.Firebase
{
    internal sealed class FileStorage : IFileStorage
    {
        private readonly FirebaseSettings _settings;
        private readonly StorageClient _storageClient;

        public FileStorage(FirebaseSettings settings, StorageClient storageClient)
        {
            _settings = settings;
            _storageClient = storageClient;
        }
        
        public async Task<MediaResult> SaveFileAsync(string filename, string contentType, Stream fileStream,
            CancellationToken cancellationToken = default)
        {
            var result = await _storageClient.UploadObjectAsync(_settings.BucketName, filename, contentType, fileStream,
                cancellationToken: cancellationToken);
            if (fileStream != null)
            {
                await fileStream.DisposeAsync();
            }
            return new MediaResult()
            {
                Id = result.Name,
                ContentType = contentType
            };
        }

        public async Task<bool> DeleteFileAsync(string filename, CancellationToken cancellationToken = default)
        {
            try
            {
                await _storageClient.DeleteObjectAsync(_settings.BucketName, filename,
                    cancellationToken: cancellationToken);
                return true;
            }
            catch (GoogleApiException ex)
            {
                if (ex.HttpStatusCode == HttpStatusCode.NotFound)
                {
                    return false;
                }

                throw;
            }
        }
    }
}